### DISCUSSION TITLE
Code cleanup

### DISCUSSION THREAD

**Discussion by lppedd** [JaneWardSandy/eclipse-pde-partial-idea#84] [General]
So, I've been browsing the code under the `config` package, while also trying to understand the logical structure of the Target Platform definition.  

I think a major problem for me currently is code readability.
I see the usage of `let`, `apply`, `also`, in chains too, is pretty high, and most files contains multiple classes.
I also see an high usage of properties (`get()`/`set()`).

Altough I know these Kotlin features are marketed as cool things, I highly recommend not using them, or just relegate their usage to small scopes. A few examples:

`?.` (and chains of them) > most of the times it means there is a logical problem in the code
`?:` > In most cases it's a lot more readable just using the classic `if (... != null) ... else ...`
`get()`/`set()` > it's cool to call the underlying method by just referencing the property, but for a reader it's difficult to understand if it's actually a simple instance variable or if it performs complex stuff. JavaBean style is still the most readable after 20 years.
`forEach` > if used in a call chain, it increase the indentation level without a reason and makes you use the `it` variable, which carries no meaning. Go for the old school enhanced `for`

If you look at my plugins repos you'll find that I still code as if I'm using Java 95% of the time. I use Kotlin stuff only when it makes the code smaller but easier to grasp.

Just my 2cents, not actually a critic.

**Comment by JaneWardSandy** [comment]
Yes, in the recent revisions, I have clearly felt that the code is not easy to read, and it is becoming more and more difficult to add logic to it.

When I first created this plug-in project, I mainly wanted to solve the problem that IDEA could not debug Teamcenter RAC (based on Eclipse RCP), and many codes were written at will.
Later, I wanted to be able to prompt and manage plugin dependencies, and added the file traversal of the target platform and the check of MANIFEST.MF, so far the code is still readable.
After changing the project name and open source to Github, various functions and requirements began to be added. However, the structure of the project and my understanding of Gradle's multi-module management still do not have a clear direction. Therefore, the code is always directly added and modified on the original basis. Change it, and then it's out of control 😂

The original plan was to wait for me to finish my work during this period, and then consider opening the v2.0 version to rewrite the entire plugin. Before that, there was still a lot of theoretical knowledge of software engineering to learn, and I was lazy. 🫠

In the planning and coding of the future v2.x version, I hope you can give me some pointers. I will also read the relevant code of the plug-in in your repository to see if it can enlighten me on the software structure organization of plug-in development.

Thank you so much for your help and inspiration to me all the time.

**Comment by lppedd** [comment]
> I have clearly felt that the code is not easy to read

There are days I write code in the morning, and in the afternoon I have the same feeling, so I start over again 🤣 
Or days I just stare at the monitor trying to image the final result. So don't rush things, and maybe try to draw diagrams, or simple steps to outline a process you have to implement.

> I hope you can give me some pointers

Sure. I'd say when you're building/rewriting something just open a draft PR and let's discuss how it's being done.

**Comment by JaneWardSandy** [comment]
I've been thinking about the final design of some modules these days and haven't really implemented them.
Next, what I want to achieve and why I want to write this way, I will open a PR draft to discuss it. There is always a blind spot in a person's fantasy.

**Reply by lppedd** [reply]
@JaneWardSandy didn't receive a notification, sorry. Sure! I'll be waiting for the PR draft then.
